src/masternodes/mn_checks.cpp

```diff
@@ -3939,15 +3942,21 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor {
                     return Res::Err("evm tx failed to validate %s", result.reason);
                 }
             }
+            
+            if (prevalidateResults.tx_fees == 0) {
+                return Res::Err("evm tx does not pay a fee");
+            }
 
             evm_try_queue_tx(result, evmContext, HexStr(obj.evmTx), tx.GetHash().GetByteArray(), prevalidateResults.gas_used);
             if (!result.ok) {
                 LogPrintf("[evm_try_queue_tx] failed, reason : %s\n", result.reason);
                 return Res::Err("evm tx failed to queue %s\n", result.reason);
             }
+
+            totalEvmFees += prevalidateResults.tx_fees;
```

Comment:
This is incorrect. The fees returned here isnt actually the gas fees that the sender will be paying, but the prepay gas fees. Will make a quick PR to make the FFI results clearer.

Reply:
The check will still pass if we are intending to check that the total prepay gas fees of the block is less than or equal to the total block fees actually paid (burnt + priority fees).

Reply:
The miner code also uses tx_fees from the evm_try_validate_raw_tx result to check if one TX pays more than another. Sounds like that will also be the incorrect behavior. Can you also refactor to include the total fees for us here and in the miner code?

Reply:
`const auto& gasFees = evmTXFees.at(evmKey);
                        if (txResult.tx_fees > gasFees) {`

If you are referring to tx fees being added to the ordered map, then this pipeline is correct. On the ethereum side, txs with the same nonce will be replaced in the mempool with a new tx with higher prepay gas fees.

Reply:
the tx_fees used in this PR are all incorrect though. evm_try_validate_tx and evm_try_prevalidate_tx do not return the actual gas fees that will be used in the tx.

Reply:
Thanks for confirming that things are correct on the miner side.

Reply:
Resolved in 74e07a2